app/src/main/kotlin/deliveryfee/App.kt

```diff
@@ -0,0 +1,179 @@
+
+
+@Serializable
+data class DeliveryRequest(
+    val cart_value: Int, // cent
+    val delivery_distance: Int, //meter
+    val number_of_items: Int,
+    val time: String // UTC
+)
+
+@Serializable
+data class DeliveryResponse(
+    val delivery_fee: Int
+)
+
+
+fun main() {
+    embeddedServer(Netty, port = 8080) {
+        install(ContentNegotiation) {
+            json(Json { ignoreUnknownKeys = true })
+        }
+
+        routing {
+            post("/delivery-fee") {
+                try {
+                    val request = call.receive<DeliveryRequest>()
+                    val deliveryFee:Int = calculateDeliveryFee(request)
+                    call.respond(DeliveryResponse(deliveryFee))
+                } catch (e: Exception) {
+                    call.respond(HttpStatusCode.InternalServerError, "Internal Server Error\nDetails: ${e}")
+                    e.printStackTrace()
+                    println("something went wrong!")
+                }
+            }
+        }
+    }.start(wait = true)
+    println("App is running!<3")
+}
+
+
+fun calculateDeliveryFee(request: DeliveryRequest): Int {
+    try{
+        val cartValue = request.cart_value
+        val deliveryDistance = request.delivery_distance
+        val numberOfItems = request.number_of_items
+        val deliveryTime = OffsetDateTime.parse(request.time)
+
+        //100€ Free delivery?
+        if (cartValue >= 10000) {
+            val deliveryFee:Int = 0
+        return  deliveryFee
+        }
+
+        val smallOrderSurcharge: Int = calculateOrderSurcharge(cartValue)
+        val distanceFee: Int = calculateDistanceFee(deliveryDistance)
+        val itemSurcharge: Int = calculateItemSurcharge(numberOfItems)
+        val bulkFee: Int = if (numberOfItems > 12) 120 else 0
+
+        var deliveryFee:Int = distanceFee + itemSurcharge + bulkFee + smallOrderSurcharge
+
+        // Rush Hour?
+        if (isRushHour(deliveryTime)) {
+            val deliveryFee: Int = calculateRushHourFee(deliveryFee)
+            return deliveryFee
+        }
+        
+        // cap 15 euro
+        if(deliveryFee >= 1500){deliveryFee = 1500}
+        return deliveryFee
+
+        }catch (e: Exception) {
+            e.printStackTrace()
+            throw e
+            println("err: main") 
+        }
+}
+
+
+fun calculateOrderSurcharge(cartValue: Int): Int{
+    try{
+        val surcharges: Int
+        surcharges = if (cartValue < 1000) {
+            1000 - cartValue
+        } else 0
+
+    return surcharges
+
+    }catch (e: Exception) {
+        e.printStackTrace()
+        throw e
+        println("err: calculateOrderSurcharge") 
+    }
+}
+
+
+fun calculateDistanceFee(distance: Int): Int {
+    try{
+        val baseDistanceFee = 200
+        val additionalDistance: Int = distance - 1000
+        val additionalDistanceFee: Int = if (additionalDistance > 0) {
+            100 * (additionalDistance / 500) + if (additionalDistance % 500 != 0) 100 else 0
```

Comment:
@dutch3883 Thanks! yeah, I also thought that I should care about rounding error, but the assignment says the return type should be "integer" so I can't use it as it return "Double".

or is there any way to avoid the "type" issue?

Reply:
I think toInt() would be the answer to my previous question. is it correct? 

ceil(additionalDistance / 500.0) * 100).toInt()

Reply:
If its correct, why do I need to use the "ceil" only in the line? I wonder what's the difference between the line and the other calculations.🤔_

Reply:
> I think toInt() would be the answer to my previous question. is it correct?
That is right. 
> why do I need to use the "ceil" only in the line?
to explain the calculation, suppose the additionalDistace is 1050
additionalDistance/500.0 = 2.01 
Math.ceil(additionalDistance/500.0) = 3.00 Double 
Math.ceil(additionalDistance/500.0).toInt = 3 Int
(additionalDistance/500.0).toInt = 2

Reply:
Ahh got it, thanks!